q2_assembly/bowtie2/mapping.py

```diff
@@ -0,0 +1,146 @@
+    """
+    base_cmd = ['bowtie2']
+    base_cmd.extend(common_args)
+
+    with tempfile.TemporaryDirectory() as tmp:
+        bam_result = os.path.join(tmp, 'alignment.bam')
+        cmd = deepcopy(base_cmd)
+        cmd.extend([
+            '-x', sample_inputs['index'],
+        ])
+
+        if paired:
+            cmd.extend(['-1', sample_inputs['fwd'],
+                        '-2', sample_inputs['rev']])
+        else:
+            cmd.extend(['-U', sample_inputs['fwd']])
+
+        try:
+            run_commands_with_pipe(
+                cmd1=cmd,
+                cmd2=['samtools', 'view', '-bS', '-o', bam_result],
+                verbose=True)
+        except subprocess.CalledProcessError as e:
+            raise Exception('An error was encountered while running Bowtie2, '
+                            f'(return code {e.returncode}), please inspect '
+                            'stdout and stderr to learn more.')
+
+        shutil.move(bam_result,
+                    os.path.join(result_fp, f'{sample_name}_alignment.bam'))
+
+        # TODO: introduce support for the stats generated by bowtie
+        #  (--met flag) and create a visualisation
+
+
+def _gather_sample_data(
+        indexed_contigs: Bowtie2IndexDirFmt, reads_manifest: pd.DataFrame,
+        paired: bool
+):
+    """Collects reads and indices for all the samples.
+
+    Args:
+        indexed_contigs (Bowtie2IndexDirFmt): Indices for all contigs.
+        reads_manifest (pd.DataFrame): Manifest with forward and
+            reverse (if paired) reads.
+        paired (bool): Indicates whether reads are paired-end.
+
+    Returns:
+        full_set (dict): Dictionary with read and index information per sample.
+    """
+    full_set = {}
+    for samp in list(reads_manifest.index):
+        full_set[samp] = {
+            'fwd': reads_manifest.loc[samp, 'forward'],
+            'rev': reads_manifest.loc[samp, 'reverse'] if paired else None
+        }
+    for samp in full_set.keys():
+        indpath = os.path.join(str(indexed_contigs), samp)
+        if os.path.exists(indpath):
+            full_set[samp]['index'] = os.path.join(indpath, 'index')
+        else:
+            raise Exception(f'Index files missing for sample {samp}. '
+                            f'Please check your input.')
+    return full_set
+
+
+def map_reads_to_contigs(
+        indexed_contigs: Bowtie2IndexDirFmt,
+        reads: Union[SingleLanePerSamplePairedEndFastqDirFmt,
+                     SingleLanePerSampleSingleEndFastqDirFmt],
+        skip: int = None, qupto: int = None, trim5: int = None,
+        trim3: int = None, trim_to: str = None, phred33: bool = False,
+        phred64: bool = False, very_fast: bool = False, fast: bool = False,
+        sensitive: bool = False, very_sensitive: bool = False,
+        very_fast_local: bool = False, fast_local: bool = False,
+        sensitive_local: bool = False, very_sensitive_local: bool = False,
```

Comment:
would it be a simpler API to have a set of choices for these? Such as `speed`, `sensitivity` and `local` (I guess that would stay `bool`)? Then the user can compose instead of having all of these (I assume mutually exclusive) arguments.

Reply:
yup, good point! I'm also all for the solution mentioned above - never really liked having all those flags exposed here... will refactor!